src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,453 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <cuda.h>
+#include <cupti.h>
+#include <iostream>
+#include <map>
+#include <pthread.h>
+#include <exception>
+
+#define BOOST_SPIRIT_THREADSAFE
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include <spdlog/spdlog.h>
+#include <sys/inotify.h>
+#include <sys/time.h>
+
+#define STDCALL
+inline const std::string FAULT_INJECTOR_CONFIG_PATH{"FAULT_INJECTOR_CONFIG_PATH"};
+const char *configFilePath = std::getenv(FAULT_INJECTOR_CONFIG_PATH.c_str());
+
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+#define CUPTI_CALL(call)                                                       \
+do {                                                                           \
+    CUptiResult _status = call;                                                \
+    if (_status != CUPTI_SUCCESS) {                                            \
+        const char *errstr;                                                    \
+        cuptiGetResultString(_status, &errstr);                                \
+        spdlog::error("function {} failed with error {}", #call, errstr);      \
+    }                                                                          \
+} while (0)
+
+#define PTHREAD_CALL(call)                                                         \
```

Comment:
Not a deal breaker, but I'm curious why we went with a macro, when we had lambdas at our disposal (`CMAKE_CXX_STANDARD 17`).
I'm guessing it's because the call site is simplified if we use a macro?

Reply:
More than anything else, this is a reflection of my prejudice against macros. Your code looks correct, at least to my eye.

Reply:
If you want to print the crash line using `__LINE__` native macro then a new macro is necessary. Otherwise either macro or function will work. But a function is preferred.